File: src/main/java/io/sc3/plethora/integration/computercraft/TurtleFakePlayerProvider.java
package io.sc3.plethora.integration.computercraft;

import com.mojang.authlib.GameProfile;
import dan200.computercraft.api.turtle.ITurtleAccess;
import dan200.computercraft.shared.turtle.TurtleUtil;
import dan200.computercraft.shared.util.DirectionUtil;
import dan200.computercraft.shared.util.InventoryUtil;
import dan200.computercraft.shared.util.WorldUtil;
import net.minecraft.entity.EquipmentSlot;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.inventory.Inventory;
import net.minecraft.item.ItemStack;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.Hand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import io.sc3.plethora.api.IPlayerOwnable;
import io.sc3.plethora.gameplay.PlethoraFakePlayer;

import java.util.WeakHashMap;

public class TurtleFakePlayerProvider {
    private static final WeakHashMap<ITurtleAccess, PlethoraFakePlayer> registeredPlayers = new WeakHashMap<>();

    public static PlethoraFakePlayer getPlayer(ITurtleAccess entity, IPlayerOwnable ownable) {
        return getPlayer(entity, ownable == null ? null : ownable.getOwningProfile());
    }

    public static PlethoraFakePlayer getPlayer(ITurtleAccess turtle, GameProfile profile) {
        PlethoraFakePlayer fake = registeredPlayers.get(turtle);
        if (fake == null) {
            fake = new PlethoraFakePlayer((ServerWorld) turtle.getLevel(), null, profile);
            registeredPlayers.put(turtle, fake);
        }

        return fake;
    }

    public static void load(PlethoraFakePlayer player, ITurtleAccess turtle, Direction direction) {
        player.setWorld((ServerWorld) turtle.getLevel());

        BlockPos position = turtle.getPosition();
        player.updatePositionAndAngles(
            position.getX() + 0.5 + 0.51 * direction.getOffsetX(),
            position.getY() + 0.5 + 0.51 * direction.getOffsetY(),
            position.getZ() + 0.5 + 0.51 * direction.getOffsetZ(),
            (direction.getAxis() != Direction.Axis.Y ? direction : turtle.getDirection()).asRotation(),
            direction.getAxis() != Direction.Axis.Y ? 0 : DirectionUtil.toPitchAngle(direction)
        );
        player.setHeadYaw(player.getYaw());

        player.setSneaking(false);

        PlayerInventory playerInv = player.getInventory();
        playerInv.selectedSlot = 0;

        // Copy primary items into player inventory and empty the rest
        Inventory turtleInv = turtle.getInventory();
        int size = turtleInv.size();
        int largerSize = playerInv.size();
        playerInv.selectedSlot = turtle.getSelectedSlot();
        for (int i = 0; i < size; i++) {
            playerInv.setStack(i, turtleInv.getStack(i));
        }
        for (int i = size; i < largerSize; i++) {
            playerInv.setStack(i, ItemStack.EMPTY);
        }

        playerInv.markDirty();

        // Add properties
        ItemStack activeStack = player.getStackInHand(Hand.MAIN_HAND);
        if (!activeStack.isEmpty()) {
            player.getAttributes().addTemporaryModifiers(activeStack.getAttributeModifiers(EquipmentSlot.MAINHAND));
        }
    }

    public static void unload(PlethoraFakePlayer player, ITurtleAccess turtle) {
        PlayerInventory playerInv = player.getInventory();
        playerInv.selectedSlot = 0;

        // Remove properties
        ItemStack activeStack = player.getStackInHand(Hand.MAIN_HAND);
        if (!activeStack.isEmpty()) {
            player.getAttributes().removeModifiers(activeStack.getAttributeModifiers(EquipmentSlot.MAINHAND));
        }

        // Copy primary items into turtle playerInv and then insert/drop the rest
        Inventory turtleInv = turtle.getInventory();
        int size = turtleInv.size();
        int largerSize = playerInv.size();
        playerInv.selectedSlot = turtle.getSelectedSlot();
        for (int i = 0; i < size; i++) {
            turtleInv.setStack(i, playerInv.getStack(i));
            playerInv.setStack(i, ItemStack.EMPTY);
        }

        for (int i = size; i < largerSize; i++) {
          TurtleUtil.storeItemOrDrop(turtle, playerInv.getStack(i));

            playerInv.setStack(i, ItemStack.EMPTY);
        }

        playerInv.markDirty();
    }
}

File: src/main/java/io/sc3/plethora/integration/computercraft/TurtleFakePlayerProvider.kt
package io.sc3.plethora.integration.computercraft

import dan200.computercraft.api.turtle.ITurtleAccess
import dan200.computercraft.shared.turtle.TurtleUtil
import dan200.computercraft.shared.util.DirectionUtil
import io.sc3.plethora.api.IPlayerOwnable
import io.sc3.plethora.gameplay.PlethoraFakePlayer
import net.minecraft.entity.EquipmentSlot
import net.minecraft.item.ItemStack
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.Hand
import net.minecraft.util.math.Direction
import java.util.*

object TurtleFakePlayerProvider {
  private val registeredPlayers = WeakHashMap<ITurtleAccess, PlethoraFakePlayer>()

  @JvmStatic
  fun getPlayer(entity: ITurtleAccess, ownable: IPlayerOwnable?): PlethoraFakePlayer =
    registeredPlayers.computeIfAbsent(entity) {
      PlethoraFakePlayer(entity.level as ServerWorld, null, ownable?.owningProfile)
    }

  @JvmStatic
  fun load(player: PlethoraFakePlayer, turtle: ITurtleAccess, dir: Direction) {
    player.setWorld(turtle.level as ServerWorld)

    val pos = turtle.position
    player.updatePositionAndAngles(
      pos.x + 0.5 + 0.51 * dir.offsetX,
      pos.y + 0.5 + 0.51 * dir.offsetY,
      pos.z + 0.5 + 0.51 * dir.offsetZ,
      (if (dir.axis !== Direction.Axis.Y) dir else turtle.direction).asRotation(),
      if (dir.axis !== Direction.Axis.Y) 0f else DirectionUtil.toPitchAngle(dir)
    )
    player.setHeadYaw(player.yaw)

    player.isSneaking = false

    val playerInv = player.inventory
    playerInv.selectedSlot = 0

    // Copy primary items into player inventory and empty the rest
    val turtleInv = turtle.inventory
    val size = turtleInv.size()
    val largerSize = playerInv.size()
    playerInv.selectedSlot = turtle.selectedSlot
    for (i in 0 until size) {
      playerInv.setStack(i, turtleInv.getStack(i))
    }
    for (i in size until largerSize) {
      playerInv.setStack(i, ItemStack.EMPTY)
    }

    playerInv.markDirty()

    // Add properties
    val activeStack = player.getStackInHand(Hand.MAIN_HAND)
    if (!activeStack.isEmpty) {
      player.attributes.addTemporaryModifiers(activeStack.getAttributeModifiers(EquipmentSlot.MAINHAND))
    }
  }

  @JvmStatic
  fun unload(player: PlethoraFakePlayer, turtle: ITurtleAccess) {
    val playerInv = player.inventory
    playerInv.selectedSlot = 0

    // Remove properties
    val activeStack = player.getStackInHand(Hand.MAIN_HAND)
    if (!activeStack.isEmpty) {
      player.attributes.removeModifiers(activeStack.getAttributeModifiers(EquipmentSlot.MAINHAND))
    }

    // Copy primary items into turtle playerInv and then insert/drop the rest
    val turtleInv = turtle.inventory
    val size = turtleInv.size()
    val largerSize = playerInv.size()
    playerInv.selectedSlot = turtle.selectedSlot
    for (i in 0 until size) {
      turtleInv.setStack(i, playerInv.getStack(i))
      playerInv.setStack(i, ItemStack.EMPTY)
    }

    for (i in size until largerSize) {
      TurtleUtil.storeItemOrDrop(turtle, playerInv.getStack(i))
      playerInv.setStack(i, ItemStack.EMPTY)
    }

    playerInv.markDirty()
    turtleInv.markDirty()
  }
}
